doc: change the parameter of virtio-net

weichenli-intel opened the pull request:
Remove device name requirement for “tap” and "vmnet", change the
parameter format like:
"-s 4,virtio-net,tap/vmnet=dev_name".
change mac_seed to virtio-net sub-parameter
As a parameter of acrn-dm,the mac_seed is only used for virtio-net.
So this patch change it to the sub-parameter of virtio-net.
The final format is:
-s 4,virtio-net,tap/vmnet=dev_name,mac_seed=XXXX

Tracked-On: #6690
Signed-off-by: Chenli Wei <chenli.wei@linux.intel.com>

acrnsi commented:
Can one of the admins verify this patch?

gvancuts commented on doc/developer-guides/hld/hostbridge-virt-hld.rst:
```diff
@@ -20,11 +20,10 @@ There is PCI host bridge emulation in DM. The bus hierarchy is determined by ``a
         -s 6,virtio-hyper_dmabuf \
         -s 3,virtio-blk,/home/acrn/UserVM.img \
-        -s 4,virtio-net,tap0 \
+        -s 4,virtio-net,tap=0,mac_seed=$mac_seed \
```
That means the tap device would be called `0`, which is a rather odd name for it, let's keep it at `tap0`, i.e.:
```suggestion
        -s 4,virtio-net,tap=tap0,mac_seed=$mac_seed \
```

weichenli-intel replied:
Sorry, Failed to change to "tap=tap0".
The virtio-net module will use snprintf to merge the type and name to the full device name.
eg.  type is tap, name is LaaG, the full device name is tap_LaaG.
If the tap0 is a confused name, I think it's better to use the tap=LaaG to show how to set virtio-net.

gvancuts replied:
Does that mean that the user has to create a TAP device called `tap_XXX` and then pass `tap=XXX` as an argument to `acrn-dm`??

weichenli-intel replied:
We have not change the create device logic.
So if device has been created and name is tap_XXX, the type is "tap".

gvancuts commented on doc/developer-guides/hld/hv-dev-passthrough.rst:
```diff
@@ -436,7 +436,7 @@ passthrough and enables PTM on it:
    acrn-dm -A -m $mem_size -s 0:0,hostbridge \
       -s 3,virtio-blk,user-vm-test.img \
-      -s 4,virtio-net,tap0 \
+      -s 4,virtio-net,tap=0 \
```
```suggestion
      -s 4,virtio-net,tap=0 \
```
```suggestion
      -s 4,virtio-net,tap=tap0 \
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

gvancuts commented on doc/developer-guides/hld/hv-dev-passthrough.rst:
```diff
@@ -506,7 +506,7 @@ If supported, follow these steps to enable PTM in the post-launched VM:
       $ acrn-dm -A -m $mem_size -s 0:0,hostbridge \
           -s 3,virtio-blk,user-vm-test.img \
-          -s 4,virtio-net,tap0 \
+          -s 4,virtio-net,tap=0 \
```
```suggestion
          -s 4,virtio-net,tap=tap0 \
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

weichenli-intel replied:
change to tap=tap0

gvancuts commented on doc/developer-guides/hld/virtio-net.rst:
```diff
@@ -484,7 +484,7 @@ optional):
 .. code-block:: none
 
-    -s 4,virtio-net,<tap_name>,[mac=<XX:XX:XX:XX:XX:XX>]
+    -s 4,virtio-net,<tap=name>,[mac=<XX:XX:XX:XX:XX:XX>]
```
I think this is a better representation of the format:
```suggestion
    -s 4,virtio-net,tap=<name>,[mac=<XX:XX:XX:XX:XX:XX>]
```

weichenli-intel replied:
I think it better to follow the current format.

weichenli-intel replied:
Done

gvancuts commented on doc/developer-guides/hld/virtio-net.rst:
```diff
@@ -533,7 +533,7 @@ a PCI slot to the device model acrn-dm as shown below.
 .. code-block:: none
 
-   -s 4,virtio-net,<macvtap_name>,[mac=<XX:XX:XX:XX:XX:XX>]
+   -s 4,virtio-net,<macvtap=name>,[mac=<XX:XX:XX:XX:XX:XX>]
```
```suggestion
   -s 4,virtio-net,macvtap=<name>,[mac=<XX:XX:XX:XX:XX:XX>]
```

gvancuts replied:
Is `macvtap` a valid name by the way, I did not think so so perhaps the right line should rather be:
```suggestion
   -s 4,virtio-net,tap=<name>,[mac=<XX:XX:XX:XX:XX:XX>]
```
?

gvancuts commented on doc/tutorials/acrn_on_qemu.rst:
```diff
@@ -267,7 +267,7 @@ Bring Up User VM (L2 Guest)
       acrn-dm -A -m $mem_size -s 0:0,hostbridge \
       -s 3,virtio-blk,~/UserVM.img \
-      -s 4,virtio-net,tap0 \
+      -s 4,virtio-net,tap=0 \
```
```suggestion
      -s 4,virtio-net,tap=tap0 \
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

gvancuts replied:
We could, but what about our network scripts (the `50-*` series) that are used to set-up a `tap0` device, should we remove them from our code and let the launch scripts deal with that?

gvancuts commented on doc/tutorials/enable_ivshmem.rst:
```diff
@@ -139,7 +139,7 @@ Linux-based post-launched VMs (VM1 and VM2).
          -s 6,virtio-hyper_dmabuf \
          -s 3,virtio-blk,/home/acrn/UserVM1.img \
-         -s 4,virtio-net,tap0 \
+         -s 4,virtio-net,tap=0 \
```
```suggestion
         -s 4,virtio-net,tap=tap0 \
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

gvancuts commented on doc/tutorials/enable_ivshmem.rst:
```diff
@@ -153,7 +153,7 @@ Linux-based post-launched VMs (VM1 and VM2).
         acrn-dm -A -m $mem_size -s 0:0,hostbridge \
          -s 3,virtio-blk,/home/acrn/UserVM2.img \
-         -s 4,virtio-net,tap0 \
+         -s 4,virtio-net,tap=0 \
```
```suggestion
         -s 4,virtio-net,tap=tap0 \
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

gvancuts commented on doc/tutorials/enable_ivshmem.rst:
```diff
@@ -220,7 +220,7 @@ Linux-based VMs (VM0 is a pre-launched VM and VM2 is a post-launched VM).
         acrn-dm -A -m $mem_size -s 0:0,hostbridge \
          -s 3,virtio-blk,/home/acrn/UserVM2.img \
-         -s 4,virtio-net,tap0 \
+         -s 4,virtio-net,tap=0 \
```
```suggestion
         -s 4,virtio-net,tap=tap0 \
```

gvancuts commented on doc/tutorials/launch_uos.sh:
```diff
@@ -135,11 +135,10 @@ acrn-dm -A -m $mem_size -c $2 -s 0:0,hostbridge -s 1:0,lpc -l com1,stdio \
   -s 8,wdt-i6300esb \
   -s 3,virtio-blk,/data/$5/$5.img \
-  -s 4,virtio-net,$tap \
+  -s 4,virtio-net,tap=$6,mac_seed=$mac_seed \
```
If I read the script properly, the tap device created by the script is actually called `tap_$6`, so this is wrong, isn't it?

weichenli-intel replied:
As previous.
The type is tap and the name is the parameter $6.

gvancuts replied:
> As previous. The type is tap and the name is the parameter $6.

The script is prepending `tap_` before creating it (or checking if it exists)... look at this line: https://github.com/projectacrn/acrn-hypervisor/blob/2bcdddaf130537796e88390d413076282e4d99cd/doc/tutorials/launch_uos.sh#L46

That means the script is assuming the TAP device is `tap_$6` but the argument to `acrn-dm` uses `%6`, this will not work.

gvancuts commented on doc/tutorials/launch_uos.sh:
```diff
@@ -313,7 +312,7 @@ ACRN project
 
  acrn-dm -A -m $mem_size -c $2 -s 0:0,hostbridge -s 1:0,lpc -l com1,stdio $npk_virt\
-   -s 9,virtio-net,$tap \
+   -s 9,virtio-net,tap=$6,mac_seed=$mac_seed  \
```
Same comment here, shouldn't this be `tap_$6`?

weichenli-intel replied:
The $tap is the full device name which spliced by the pre str "tap_" and the parameter "$6".
So the type is tap and the name is $6.

gvancuts replied:
I'm a bit confused by all this, so our code makes an assumption that while the user passes a name such as `LaaG` (or any other name), the name of the actually TAP device must be `tap_LaaG`??

weichenli-intel replied:
The actually  device name is type_name which follow the old devices name format.

That means the old command  "-s 9,virtio-net,macvtap_LaaG" should change to "-s 9,virtio-net,macvtap=LaaG".

The actually devices is macvtap_LaaG.

gvancuts replied:
This is not what we should do, the previous mechanism forced the user to include `tap` in the device name so it could be properly identified. The idea is to move away from that restriction (on the name) and be able to use any name. This is why we now have to specify the type of device as well. As an example: `-s 9,virtio-net,tap=LaaG` means that User VM will connect to a TAP device called `LaaG`.

weichenli-intel replied:
I have resolved other open, let's focus on the "tap0".

The current patch have not change the create device logic.

So if user could create a device name "tab_xxx" as the current design, it's effective also.

gvancuts replied:
Same problem here

gvancuts commented on doc/tutorials/using_windows_as_user_vm.rst:
```diff
@@ -105,7 +105,7 @@ Prepare the Script to Create an Image
    acrn-dm -A -m $mem_size -s 0:0,hostbridge -s 1:0,lpc -l com1,stdio \
      -s 2,passthru,0/2/0 \
-     -s 8,virtio-net,tap0 \
+     -s 8,virtio-net,tap=0 \
```
```suggestion
     -s 8,virtio-net,tap=tap0 \
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

weichenli-intel replied:
The "dm: remove the assume of virtio-net device name"  was PR now.
The new format is tap=tap0, so we can resolved this.

gvancuts commented on doc/tutorials/using_windows_as_user_vm.rst:
```diff
@@ -266,7 +266,7 @@ Explanation for acrn-dm Popular Command Lines
   You may need to change 0/2/0 to match the bdf of the VGA controller on your platform.
 
-* ``-s 8,virtio-net,tap0``:
+* ``-s 8,virtio-net,tap=0``:
```
```suggestion
* ``-s 8,virtio-net,tap=tap0``:
```

weichenli-intel replied:
As previous.
Could we use tap=LaaG to show the virtio-net format?

weichenli-intel replied:
change to tap=tap0

gvancuts commented on doc/user-guides/acrn-dm-parameters.rst:
```diff
@@ -469,10 +456,19 @@ arguments used for configuration.  Here is a table describing these emulated dev
-       ``virtio-net,<device_name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>]``.
-       The ``mac`` address is optional, ``device_name`` is the name of the TAP (or MacVTap) device.
-       It must include the keyword ``tap``.
+       ``virtio-net,<device_type=name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>],[mac_seed=<seed_string>]``.
```
```suggestion
       ``virtio-net,<device_type>=<name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>],[mac_seed=<seed_string>]``.
```

weichenli-intel replied:
As previous.
Is better to follow the current format?

gvancuts replied:
Not following you, but this is wrong.

gvancuts replied:
Still wrong...

gvancuts commented on doc/user-guides/acrn-dm-parameters.rst:
```diff
@@ -469,10 +456,19 @@ arguments used for configuration.  Here is a table describing these emulated dev
-       The ``mac`` address is optional, ``device_name`` is the name of the TAP (or MacVTap) device.
-       It must include the keyword ``tap``.
+       ``virtio-net,<device_type=name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>],[mac_seed=<seed_string>]``.
+       The ``mac`` address is optional, ``name`` is the name of the TAP (or MacVTap) device.
```
Can you also add the various values that `device_type` can take?

weichenli-intel replied:
Update:
"device_type" is the type of the TAP (or MacVTap) device. It must include the keyword ``tap`` .

gvancuts replied:
That's not what we had discussed, there should not be any assumption about the name.

weichenli-intel replied:
We should redesign the device name or the type and name logic?

gvancuts replied:
Take a look at #6727 as well .

weichenli-intel replied:
change to tap=tap0

amyreye commented on doc/developer-guides/hld/hld-devicemodel.rst:
```diff
@@ -110,7 +109,7 @@ Here's an example showing how to run a VM with:
      -s 5,virtio-console,@pty:pty_port \
      -s 3,virtio-blk,/home/acrn/UserVM.img \
-     -s 4,virtio-net,tap_LaaG \
+     -s 4,virtio-net,tap=LaaG \
```
@weichenli-intel For my education, what does `tap=LaaG` mean? Line 102 says "Virtio-net device on PCI 00:04.0" -- is LaaG a device name and where is the LaaG device defined (i.e., how does LaaG correspond to PCI 00:04:0)?

amyreye commented on doc/developer-guides/hld/virtio-net.rst:
```diff
@@ -484,7 +484,7 @@ optional):
 .. code-block:: none
 
-    -s 4,virtio-net,<tap_name>,[mac=<XX:XX:XX:XX:XX:XX>]
+    -s 4,virtio-net,<tap=name>,[mac=<XX:XX:XX:XX:XX:XX>]
```
@weichenli-intel Need to mention mac_seed?

weichenli-intel replied:
Done

amyreye commented on doc/developer-guides/hld/virtio-net.rst:
```diff
@@ -533,7 +533,7 @@ a PCI slot to the device model acrn-dm as shown below.
 .. code-block:: none
 
-   -s 4,virtio-net,<macvtap_name>,[mac=<XX:XX:XX:XX:XX:XX>]
+   -s 4,virtio-net,<macvtap=name>,[mac=<XX:XX:XX:XX:XX:XX>]
```
@weichenli-intel Need to mention mac_seed?

weichenli-intel replied:
Yes, I have add some mention of mac_seed.

amyreye commented on doc/user-guides/acrn-dm-parameters.rst:
```diff
@@ -469,10 +456,19 @@ arguments used for configuration.  Here is a table describing these emulated dev
+
+   mac=$(cat /sys/class/net/e*/address)
+   seed_string=${mac:9:8}-${vm_name}
 
```
@weichenli-intel Please indent lines 468-471 under line 466.

weichenli-intel replied:
I will check it.

amyreye commented on doc/user-guides/acrn-dm-parameters.rst:
```diff
@@ -469,10 +456,19 @@ arguments used for configuration.  Here is a table describing these emulated dev
-       It must include the keyword ``tap``.
+       ``virtio-net,<device_type=name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>],[mac_seed=<seed_string>]``.
+       The ``mac`` address is optional, ``name`` is the name of the TAP (or MacVTap) device.
        ``vhost`` specifies vhost backend, otherwise the VBSU backend is used.
```
@weichenli-intel The commit message says "The final format is: `-s 4,virtio-net,tap/vmnet=dev_name,mac_seed=XXXX`". However, I don't see vmnet mentioned anywhere in the changes. Need to add description of vmnet?

weichenli-intel replied:
I have search current doc.
It's strange there was no mentioned of vmnet.
So I need check why and whether we should add it.

weichenli-intel replied:
Let's keep the silent of vmnet.

gvancuts commented on doc/developer-guides/hld/virtio-net.rst:
```diff
@@ -533,7 +533,7 @@ a PCI slot to the device model acrn-dm as shown below.
 .. code-block:: none
 
-   -s 4,virtio-net,<macvtap_name>,[mac=<XX:XX:XX:XX:XX:XX>]
+   -s 4,virtio-net,macvtap=<name>,[mac=<XX:XX:XX:XX:XX:XX>]
```
I thought the only valid options here were `tap` and `vmnet`, have I missed something?

weichenli-intel replied:
As we have talk, we should remove the macvtap from doc.

dbkinder commented:
@weichenli-intel Looks like there are merge conflicts to resolve now... need to rebase against master.  And I think there are still some unresolved comments from @gvancuts to deal with...

dbkinder commented:
@NanlinXie @gvancuts @weichenli-intel This PR seems to have stalled.  Let's get an update at our next doc meeting.

gvancuts commented on doc/developer-guides/hld/virtio-net.rst:
```diff
@@ -525,15 +525,14 @@ Create a MacVTap interface in the Service VM as shown here:
 .. code-block:: none
 
-   -s 4,virtio-net,<macvtap_name>,[mac=<XX:XX:XX:XX:XX:XX>]
+   -s 4,virtio-net,tap=<macvtap0>,[mac=<XX:XX:XX:XX:XX:XX>]
```
```suggestion
   -s 4,virtio-net,tap=macvtap0,[mac=<XX:XX:XX:XX:XX:XX>]
```

gvancuts commented on doc/user-guides/acrn-dm-parameters.rst:
```diff
@@ -467,10 +454,19 @@ arguments used for configuration.  Here is a table describing these emulated dev
-       ``virtio-net,<device_name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>]``.
-       The ``mac`` address is optional, ``device_name`` is the name of the TAP (or MacVTap) device.
-       It must include the keyword ``tap``.
+       ``virtio-net,<device_type=name>,[vhost],[mac=<XX:XX:XX:XX:XX:XX>],[mac_seed=<seed_string>]``.
```
I'm assuming the use of `mac` and `mac_seed` is mutually exclusive, isn't it?

If so, we should make this clear in the doc,

weichenli-intel commented:
"I'm assuming the use of mac and mac_seed is mutually exclusive, isn't it?
If so, we should make this clear in the doc,"

——mac priority higher then mac_seed.
virtio-net will check whether set mac, if no, use mac_seed.
BTW, if user have not set mac and mac_seed, virtio-net will use solt and func info to generate mac.

gvancuts commented:
> ——mac priority higher then mac_seed. virtio-net will check whether set mac, if no, use mac_seed. BTW, if user have not set mac and mac_seed, virtio-net will use solt and func info to generate mac.

Can we add that in the explanation (maybe as a note). I think this is useful for the user to know.

weichenli-intel commented:
> > ——mac priority higher then mac_seed. virtio-net will check whether set mac, if no, use mac_seed. BTW, if user have not set mac and mac_seed, virtio-net will use solt and func info to generate mac.
> 
> Can we add that in the explanation (maybe as a note). I think this is useful for the user to know.

OK， I will update the doc.

gvancuts reviewed (approved):
Thanks, just update the commit message to remove the reference to `vmnet` and this is good to go from my perspective.

@dbkinder may have further feedback on the text.

wenlingz commented:
make doc fail:
New errors/warnings found, please fix them:
==============================================

/home/acrn/workspace/acrn-hypervisor-build-doc_248/acrn-hypervisor/build/doc/rst/user-guides/acrn-dm-parameters.rst:466: WARNING: Content block expected for the "note" directive; none found.
/home/acrn/workspace/acrn-hypervisor-build-doc_248/acrn-hypervisor/build/doc/rst/user-guides/acrn-dm-parameters.rst:467: WARNING: Explicit markup ends without a blank line; unexpected unindent.

make[1]: *** [Makefile:64: html] Error 1
make[1]: Leaving directory '/home/acrn/workspace/acrn-hypervisor-build-doc_248/acrn-hypervisor/doc'
make: *** [Makefile:147: doc] Error 2

gvancuts commented:
@weichenli-intel , I have just fixed the `make doc` issue, updated the commit message and made some more changes to the plain text. I have force-pushed these changes to your remote branch so they are now the latest in this PR... hope you don't mind ;-)

weichenli-intel commented:
> @weichenli-intel , I have just fixed the `make doc` issue, updated the commit message and made some more changes to the plain text. I have force-pushed these changes to your remote branch so they are now the latest in this PR... hope you don't mind ;-)

Very Thanks

dbkinder commented:
@NanlinXie @wenlingz @weichenli-intel Sigh.  I tried using the GitHub UI to resolve the merge conflict, but it didn't appear to work.  The build doc CI test failed, but I can't see what the failure is.  Can you help figure out what's wrong...
